pkg/codegen/go/gen_program_expressions.go

```diff
@@ -235,7 +238,34 @@ func (g *generator) GenFunctionCallExpression(w io.Writer, expr *model.FunctionC
 		case *model.ScopeTraversalExpression:
 			g.genScopeTraversalExpression(w, arg, expr.Type())
 		default:
-			g.Fgenf(w, "%.v", expr.Args[0])
+			// Add a cast to the type we expect if needed
+			if originalTo.AssignableFrom(from.Type()) && (isOutput == isFromOutput) {
+				g.Fgenf(w, "%.v", from)
+			} else {
+				typeName := g.argumentTypeName(from, to, isOutput)
+				// IDOutput has a special case where it can be converted to a string
+				var isID bool
+				if expr, ok := from.(*model.ScopeTraversalExpression); ok {
+					last := expr.Traversal[len(expr.Traversal)-1]
+					if attr, ok := last.(hcl.TraverseAttr); ok && attr.Name == "id" {
+						isID = true
+					}
+				}
+				if expr, ok := from.(*model.RelativeTraversalExpression); ok {
```

Comment:
Assuming these clauses are mutually exclusive, is this better as an `else if` or a type `switch`?

Reply:
Yeh can switch it